python/tgn.py

```diff
@@ -0,0 +1,1171 @@
+
+    num_eval_edges = len(eval_edges)
+    num_eval_batches = ceil(num_eval_edges / batch_size)
+    assert batch_size > 0
+
+    for epoch in range(num_epochs):
+
+        # update global epoch counter
+        update_epoch_counter()
+
+        # initialize container for current epochs where we save records for
+        # accuracy results and batch processing time
+        initialize_results_per_epoch(get_current_epoch())
+
+        # on every epoch training tgn should have clean state,
+        # although we update parameters, memory should be empty,
+        # temporal neighborhood should be empty and message store should be empty
+        # because if it isn't we have problem with information leakage from future
+        # to current training samples
+        query_module_tgn.tgn.init_memory()
+        query_module_tgn.tgn.init_temporal_neighborhood()
+        query_module_tgn.tgn.init_message_store()
+
+        query_module_tgn.all_edges = set()
+        query_module_tgn.m_loss = []
+
+        reset_tgn_batch(batch_size=batch_size)
+
+        # when we update here tgn_mode, later when we call process_batch_self_supervised
+        # it will change tgn mode to .train() so no worries
+        query_module_tgn.tgn_mode = TGNMode.Train
+
+        for i in range(num_train_batches):
+            # sample edges we need
+            start_index_train_batch = i * batch_size
+            end_index_train_batch = min((i + 1) * batch_size, num_train_edges)
+            current_edges_batch = train_edges[
+                start_index_train_batch:end_index_train_batch
+            ]
+
+            # prepare batch
+            parse_mgp_edges_into_tgn_batch(current_edges_batch)
+            batch_result_record = process_epoch_batch()
+            append_batch_record_curr_epoch(get_current_epoch(), batch_result_record)
+
+            # reset for next batch
+            reset_tgn_batch(batch_size=batch_size)
+
+        # here we need to change mode to eval
+        # also later when we call process_batch_self_supervised
+        # it will change tgn mode to .eval()
+        query_module_tgn.tgn_mode = TGNMode.Eval
+        for i in range(num_eval_batches):
+            # sample edges we need
+            start_index_eval_batch = i * batch_size
+            end_index_eval_batch = min((i + 1) * batch_size, num_eval_edges)
+            current_edges_batch = eval_edges[
+                start_index_eval_batch:end_index_eval_batch
+            ]
+            # prepare batch
+            parse_mgp_edges_into_tgn_batch(current_edges_batch)
+            batch_result_record = process_epoch_batch()
+            append_batch_record_curr_epoch(get_current_epoch(), batch_result_record)
+
+            # reset for next batch
+            reset_tgn_batch(batch_size=batch_size)
+
+
+#####################################################
+
+# all available read_procs
+
+#####################################################
+
+
+@mgp.read_proc
+def process_epochs(
```

Comment:
Rename to `train`. Implicitely set mode to `eval`. Create an eval function to fetch accuracy.